python/pyarrow/tests/test_parquet.py

```diff
@@ -2871,6 +2859,71 @@ def test_read_column_invalid_index():
             f.reader.read_column(index)
 
 
+def test_direct_read_dictionary():
+    # ARROW-3325
+    repeats = 10
+    nunique = 5
+
+    data = [
+        [tm.rands(10) for i in range(nunique)] * repeats,
+
+    ]
+    table = pa.table(data, names=['f0'])
+
+    bio = pa.BufferOutputStream()
+    pq.write_table(table, bio)
+    contents = bio.getvalue()
+
+    result = pq.read_table(pa.BufferReader(contents),
+                           read_dictionary=['f0'])
+
+    # Compute dictionary-encoded subfield
+    expected = pa.table([table[0].dictionary_encode()], names=['f0'])
+    assert result.equals(expected)
+
+
+def test_dataset_read_dictionary(tempdir):
+    path = tempdir / "ARROW-3325-dataset"
+    data = [
+        [tm.rands(10) for i in range(5)] * 10,
+    ]
+    table = pa.table(data, names=['f0'])
+    pq.write_to_dataset(table, root_path=str(path))
+    result = pq.ParquetDataset(path, read_dictionary=['f0']).read()
+    expected = pa.table([table[0].dictionary_encode()], names=['f0'])
+    assert result.equals(expected)
```

Comment:
Does this already work for a partitioned dataset with multiple parquet files where a the different files might have different set of unique values?

Reply:
Each chunk in the table will have a different dictionary, yeah. So there shouldn't be any problem

Reply:
I'll expand the unit test to check explicitly

Reply:
Done